### ISSUE TITLE
Q: projecting single time point on tangent space?

### ISSUE THREAD

**Issue by kingjr** [alexandrebarachant/pyRiemann#9]
Hi, 

I'm trying to play with your code.

I was wondering whether it would be in principle possible to project a single time point (Nchan x 1) on the tangent space. I was thinking of fitting `TangentSpace` on the complete epochs, but then assess a series of svm on each time point separately in the tangent space. Although I'm not sure this makes sense at all..

Thanks!

**Comment by alexandrebarachant** [comment]
Hi,

It is totally fine, once you called the `fit` method, each sample is treated independently.
However the `transform` method expect a 3D tensor of covariance matrices, so if you want to call it with a single matrix, you should put it in a 3D tensor of shape `[1 x Nchann x Nchann]`

you can do this with the methos `atleast_3d` of [numpy](http://docs.scipy.org/doc/numpy/reference/generated/numpy.atleast_3d.html).

Note that it won't works if you use the parameter `tsupdate=True` of `TangentSpace` because it re-estimate the tangent space reference point, as described in section 3.3 of my [article](https://hal.archives-ouvertes.fr/file/index/docid/820475/filename/BARACHANT_Neurocomputing_ForHal.pdf).

It is the same for the method `inverse_transform` (used to retro-project a vector in the manifold). You can pass a single vector, but the function expect a 2d array.

**Comment by kingjr** [comment]
Thanks for the quick response. I'm not sure to understand your proposal, and I think it's because I didn't explained myself very well.

In case you want to do a [search light](http://martinos.org/mne/dev/auto_examples/decoding/plot_decoding_sensors.html), where the shape of X is `n_epochs` x `n_channels` x `n_times`, would it be possible / make sense to project a single time point of a a single epoch onto the tangent space?

Something like:
```
times = range(X.shape[2])
scores = np.zeros(X.shape[0], X.shape[2])
Xcov = Covariances.fit(X[train])
ts = TangentSpace()
for train, test in cv:
    ts.fit(X[train], y[train])
    for t in times:
        Xts = ts.transform(X[test][:, t])
        scores[test, t] = clf.score(Xts, y[test])
```

Thanks!

**Comment by alexandrebarachant** [comment]
I see. indeed, i did not understand your question :(

The tangent space projection is only valid for SPD (symmetric & positive-definite) matrices. So you can not project a time sample in the tangent space.

What you can do is to estimates covariances matrices using a very narrow time window, project the results in tangent space and apply the classifier.
I'm not sure you can estimate a covariance matrice using a single time sample (it depends on the estimator you use).
Using a short time window Generally require a regularized estimation of the cov mat, since they can be no longer SPD.

If you want a search light, you have to do something like that (using sklearn pipeline)

```python

window = 20 # time sample
scores = np.zeros(X.shape[0], X.shape[2])
times = range(window/2, X.shape[2] - window/2)

# create the classifier by pipelining the covariance matrix estimation (ledois-wolf),
# the tangent space projection, and a classifier
clf = make_pipeline( Covariances(estimator='lwf'), TangentSpace(), LogisticRegression())

for train, test in cv:
    clf.fit(X[train], y[train])
    for t in times:
        Xts = X[test][:, (t - window / 2):(t + window / 2)]
        scores[test, t] = clf.score(Xts, y[test])
```

Keep in mind that the length of the window will have an impact on performance, and that regularization can decrease performance too

**Comment by kingjr** [comment]
Thanks. Unfortunately this approach doesn't seem to work. Here s a [gist](https://gist.github.com/kingjr/8eeaacf9b44ea4cde169) for future references.

On a related topic, I think you should change the method `_predict_distances` into the more classic sklearn `decision_function`. This can be important for allowing continuous scoring. BTW, did you try to compare the scores of LDA vs LDA+MDM using continuous scoring approach (e.g. [auc](http://scikit-learn.org/stable/modules/generated/sklearn.metrics.roc_auc_score.html))?

**Comment by alexandrebarachant** [comment]
I see, you are actually working on evoked potential, and the 'Covariances' estimation is not very adapted.
in Addition, you are using meg data, with a very high number of sensors, and the RG can not handle it.

Here is an adapted [gist](https://gist.github.com/alexandrebarachant/54a3d0d64603d7a458de)
I subsampled the electrodes by a factor 10 (~20 elec) 
Here is the results : 
![image](https://cloud.githubusercontent.com/assets/2145718/8194263/843a849c-1448-11e5-852f-7a6fa17ed2ad.png)

I'm actually very surprised to see it working. 

If you want to achieve a better accuracy, you have to reduce the dimensionality of the signals, and use a special estimation of covariance matrices dedicated to ERP. you can do this with the XdawnCovariances estimator.
Here is the corresponding [gist](https://gist.github.com/alexandrebarachant/d7171d9af6de31ce8caa)
I will have the results tomorrow, but it should works

However, this is a pattern-matching method, meaning that you are classifying based on correlation with a pattern of reference. This is clearly not adapted to do this kind of thing.

**Comment by kingjr** [comment]
Impressive reactivity! Thanks a lot for the help and opinion. 

> I subsampled the electrodes by a factor 10 (~20 elec) 

Ok. An ICA with n_ranks components should probably loose less information.

> However, this is a pattern-matching method, meaning that you are classifying based on correlation with a pattern of reference. This is clearly not adapted to do this kind of thing.
 instead?
Ok, my intuitive rational is that using covariance estimates instead of the classic LDA across channels may be able to detect induced signals (i.e. non-phase locked signals across trials), which contains more information than evoked signals. But perhaps this is an incorrect intuition?

**Comment by alexandrebarachant** [comment]
> Impressive reactivity! Thanks a lot for the help and opinion.

I had some spare time today :)

> Ok. An ICA with n_ranks components should probably loose less information.

Sure ! Or a CSP, so you are sure your not missing some interesting components.

> Ok, my intuitive rational is that using covariance estimates instead of the classic LDA across channels may be able to detect induced signals (i.e. non-phase locked signals across trials), which contains more information than evoked signals. But perhaps this is an incorrect intuition?

Your intuition is correct. The covariance based feature allow to catch non-phase locked patterns. However for a typical ERP classification problem, the phase locked response is much more informative, and sometimes, there is no induced signals at all. Moreover, for catching this, you need multiple time samples.
For ERP classification, i use a special estimation of the covariance matrix, decribed [here](http://arxiv.org/pdf/1409.0107v1.pdf), section III.B
The strength is that it contain both information (phase locked and induced), so you get the best of the two world, with a single classifier.

Again, you need multiple time sample to take advantage of this.

**Comment by kingjr** [comment]
> Moreover, for catching this, you need multiple time samples.

Necessarily? What about if you
1. estimate the covariance of each class on the entire epoch
2. find the discriminatory eigenvectors,
3. project each time sample onto these eigenvectors separately 
4. test to which class each time sample is closest.

In principle this could capture an induced signals (although it would be dominated by an evoked one). 

I'll try to see whether your III.B can help my case. TBC

**Comment by alexandrebarachant** [comment]
This is more or less what is done here : http://mne-tools.github.io/stable/auto_examples/decoding/plot_decoding_csp_eeg.html

Induced activity are variance / frequency based activity. if you want to catch them, you need enough time point to estimates the signal power.

Anyway, i updated the first gist, tweaking a few parameters (bandpass filter, no time or channel decimation), and i got this 
![figure_1](https://cloud.githubusercontent.com/assets/2145718/8211695/e8503eb6-14e5-11e5-908a-c21ccc1e24bf.png)

Giving the time window i set, and the frequency content of the signal, i doubt we are catching an induced activity. The algorithm probably react to the peak of the ERP (maximum variance).

i also made another [gist](https://gist.github.com/alexandrebarachant/8975dc6d9dac9c07ed0a) using `ERPCovariances` with an SVD to reduce dimension of the Evoked potential before concatenation. (it's a fresh commit, you have to update pyriemann from the sources).
it gave me this results

**Comment by kingjr** [comment]
Awesome. I suspect the differences in scores may be a bit inflated by your scoring metrics. I would recommend comparing the AUC and not the accuracy, which doesn't behave so greatly for difficult cases.

Here's the comparison of your approach ('riemann') to the classic approach (StandardScaler + Logistic with probabilistic output) using an AUC score. It's probably worthwhile digging a bit this alley and test whether there's a genuine benefit. ATM, it's a bit annoying because the covariance estimates are pretty long to compute, but perhaps we could increase the threshold to converge more rapidly?

![compare](https://cloud.githubusercontent.com/assets/4881164/8218258/de3dc612-150f-11e5-85ab-273130bc36fc.png)
output of this [gist](https://gist.github.com/kingjr/5c5b3fdcd9fa1cb07e4b). See https://github.com/alexandrebarachant/pyRiemann/pull/11 to make it work.

**Comment by alexandrebarachant** [comment]
This is usually the other way around. AUC give higher results, since the accuracy can be influenced by the offset on the prediction probabilities. This especially the case for high dimentional data with a big covariate shift between training and test.

When i switch the scoring to the auc ( `scoring='roc_auc` in `cross_val_score`) i get this 

![figure_1](https://cloud.githubusercontent.com/assets/2145718/8233620/46fda0e2-15a5-11e5-80c3-ef7d7d86e9d3.png)

However, as compared to single time point classification, results are inflated since we use a bigger time window (it contains more information).

Are you sure `reshape_X` is not messing with the signal ?

If you want to speed up the computation, use the parameter `TangentSpace(metric='logeuclid')`, it should not degrades performances too much.

Finally, your use case is really borderline for the Riemannian Geometry. You are estimating covariance matrices with a really small number of samples, and we have to rely on a heavy regularisation to make it works.  In addition, i did not recommand using RG when you have a number of channel > ~60. For EEG, the noise generally degrades performances if you have too many channel. I'm quite surprised it run with 200 channels. to be honnest, i never tryed this. I guess MEG data are clean enough to make it works

**Comment by kingjr** [comment]
> This is usually the other way around. AUC give higher results, since the accuracy can be influenced by the offset on the prediction probabilities. This especially the case for high dimentional data with a big covariate shift between training and test.

I agree. What I meant regards whether the *benefit* of your Riemann's approach (over classic methods) is still visible with an AUC. I'm thinking of the scenario where your approach *only* provides a better hyperplan intercept than classic methods, in which case you'd see better accuracy but not a better AUC.

> When i switch the scoring to the auc ( scoring='roc_auc in cross_val_score) i get this 

Is this with a `decision_function` or `predict_proba`?

> Are you sure reshape_X is not messing with the signal ?

It seemed to be ok, but I didn't do an intense checkup.

> However, as compared to single time point classification, results are inflated since we use a bigger time window (it contains more information).

In the [gist](https://gist.github.com/kingjr/5c5b3fdcd9fa1cb07e4b) where I compare the classic sliding window and the Riemann's approach, the classifiers use the same input data (see `GeneralizationAcrossTime(train_time=dict(length=window))`). In the classic approach, each pair of channel x time is like an independent dimension: i.e. `X.shape = [n_trials, n_channels * n_times]`, but in the Riemann, it's reshaped into `[n_trials, n_channels, n_times]` so as to estimate the covariance across time.

**Comment by alexandrebarachant** [comment]
>  I'm thinking of the scenario where your approach only provides a better hyperplan intercept than classic methods, in which case you'd see better accuracy but not a better AUC.

To my experience, it is also the other way. RG is really sensitive to change in covariance structure that occurs naturally with time in M/EEG. As a result, the intercept can be badly calibrated, but thanks to the property of invariance of the Riemannian metric, this translate into a shift (and sometime a small rotation) in the tangent space. So AUC get generally better than other approaches, but accuracy can be a bit lower.

> Is this with a decision_function or predict_proba?

I'm not 100% sure, but i think it's with predict_proba

> It seemed to be ok, but I didn't do an intense checkup.

I had a few surprise with the reshape function in the past.

> the classifiers use the same input data

Sure, what i mean by that is that the RG make a better use of this time window. For example, the dimension of the feature space for the standard classifier increase with the window, and it can lead to more overfitting etc. The Tangent space dimension remains constant (N x (N +1))/2 with N the number of channels.

What is your goal with this ? I did not read your paper about GAT, but if the goal is to use the classification score as a metric for similar activity, we can try the same thing with a Riemannian permutations test.

**Comment by kingjr** [comment]
`metric='logeuclid'` does indeed increase sig. speed.

Here's the comparison on a larger time scale. It seems that early signals seem are better captured by the classic approach whereas late signals are better captured by the Riemann's geometry approach.

That fits with our discussion on evoked and induced?

![compare](https://cloud.githubusercontent.com/assets/4881164/8234999/673506ea-15ad-11e5-82fb-8f9e7616c1e6.png)

**Comment by kingjr** [comment]
We agree on all points.

> What is your goal with this ? I did not read your paper about GAT, but if the goal is to use the classification score as a metric for similar activity, we can try the same thing with a Riemannian permutations test.

My goal is to have the best classifier at a given time point. I have several experiment where SNR is low, and want to extract as much as I can, but still know when information rises.

**Comment by alexandrebarachant** [comment]
This is weird, we don't get the same results.
Are you evaluating AUC on the aggregated prediction from all CV partitions ? If you are interested in the greatest accuracy, you don't want to do that. The average of AUC over the partitions is a much more reliable metrics.

Also, you can increase the number of partition for the CV, and shuffle the trials.

**Comment by kingjr** [comment]
> Are you evaluating AUC on the aggregated prediction from all CV partitions ? 

Yes. Although this particular figure was actually done on a validation (one train, one test, no cross val).

> If you are interested in the greatest accuracy, you don't want to do that.

Why not?

> Also, you can increase the number of partition for the CV, and shuffle the trials.

Sure anything would work, but the folds in cross validations are not independent, so it's hard to make stats anyway. The only reliable way would be to do some permutations analyses, or use multiple subjects.

**Comment by alexandrebarachant** [comment]
> Why not?

You are training different classifiers, so their output can be differents, in terms of dynamics, but also in terms of biais. The AUC is greatly impacted by this. You can actually have a perfectly separable output, but with a different biais for each partition, and the global auc will be bad. 
This is actually fine if you want to put the emphasis on the stability of the classifier across partition, but if you are solely interested of the classification accuracy it is not optimal.

> Sure anything would work, but the folds in cross validations are not independent, so it's hard to make stats anyway

You should be able to reproduce the same resutls as the one i posted later, this is why i'm curious.

**Comment by kingjr** [comment]
> You are training different classifiers, so their output can be differents, in terms of dynamics, but also in terms of bias. The AUC is greatly impacted by this. You can actually have a perfectly separable output, but with a different bias for each partition, and the global auc will be bad.
This is actually fine if you want to put the emphasis on the stability of the classifier across partition, but if you are solely interested of the classification accuracy it is not optimal.

Thanks a lot. I actually didn't think of this; although I must say I can't think of nor simulate scenarios with randomized cross validations that lead to this bias. Also, this should be sig. diminished with recalibration (e.g. [Platt's method](https://en.wikipedia.org/wiki/Platt_scaling)) like sklearn does for SVC.predict_proba.

> You should be able to reproduce the same resutls as the one i posted later, this is why i'm curious.

It's mainly because I used a windows whom size is half yours. Here's the result with the a 4 fold CV (before it was just a validation), the same window as your, and the same scoring. Errorbars are SEM, thick line is mean on score concatenated pred (no bias seen here).

So the early versus late differences hold. If you don't see any trivial thing that explain it, I think that it's actually pretty cool :)

![compare](https://cloud.githubusercontent.com/assets/4881164/8242887/9f0e787a-15e0-11e5-9cd6-1e2fa9502e2f.png)

**Comment by kingjr** [comment]
I updated the [gist](https://gist.github.com/kingjr/5c5b3fdcd9fa1cb07e4b) if you want to try it yourself.

**Comment by alexandrebarachant** [comment]
That's great, thanks.

The second response could be a induced activity, or simply the activation due to the next stimulus (i did not look, but if the classes are somehow not completly randomized ...). The RG is also less sensitive to the jitter on the stimulations.

You can run the full GAT, i tried and see that the classifier trained on the late activity give an AUC of 0 for the data of the first response (inverted labels).

**Comment by kingjr** [comment]
> The second response could be a induced activity, or simply the activation due to the next stimulus (i did not look, but if the classes are somehow not completly randomized ...). 

Good call: `y[np.where(y[:-1]==1)[0]+1]`shows that the trials are not randomized at all! Ouch #badDesign

> You can run the full GAT, i tried and see that the classifier trained on the late activity give an AUC of 0 for the data of the first response (inverted labels).

I think for a full GAT, it's better to only have temporally precise classifiers, as the whole point is to check whether the underlying response is homogeneous across time. WDT?

**Comment by alexandrebarachant** [comment]
> I think for a full GAT, it's better to only have temporally precise classifiers, as the whole point is to check whether the underlying response is homogeneous across time. WDT?

yes, this particular ERP use case is not very adapted for the RG.
However, you should try on Motor imagery, you will be able to catch the ERD/ERS and other stuff. With a window of 200ms, you should get nice results.
